### PR TITLE
Optimize lineage eviction efficiency

### PR DESCRIPTION
**Pull request by raulchen**
=== Problem ===
Previously, when forwarding a task, we request a notification from GCS if the uncommitted lineage size is larger than `max_lineage_size` (configurable, 100 by default). The purpose of doing this is to evict these uncommitted lineage once is task's lineage is committed by a remote node.

However, this could have some perf issues in some cases. For example, if one worker forwards a large number (`N`) of tasks very quickly to an actor (this is common in streaming scenario). If the downstream actor doesn't process the tasks fast enough. We'll end up with requesting `N - max_lineage_size` notifications, which may bring significant workload to both local node manager and the GCS.

=== Solution ===
This RP resolves this issue by only request notifications for every `max_lineage_size` tasks. By doing this, we can reduce the number of requested notifications to at most `N / max_lineage_size`.

Also, it reduced the complexity of counting uncommitted lineage from `1 + 2 + ... + (N - max_lineage_size)` to `N`.

### CODE REVIEW

**Comment by raulchen** [comment]
@stephanie-wang could you please take a look? thank you!

File: src/ray/raylet/lineage_cache.h [hunk 1]
```diff
@@ -220,6 +220,8 @@ class LineageCache {
   /// Unsubscribe from notifications for a task. Returns whether the operation
   /// was successful (whether we were subscribed).
   bool UnsubscribeTask(const UniqueID &task_id);
+  /// Count the size of unsubscribed and uncommitted lineage
+  uint64_t CountUnsubscribedLineage(const UniqueID &task_id);
```
Comment (stephanie-wang) [inline src/ray/raylet/lineage_cache.h]: Can you make this `const`?

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Thanks a lot, this looks great! I left one small comment, but then we can merge this.

**Comment by robertnishihara** [comment]
Looks like there are some linting errors https://travis-ci.com/ray-project/ray/jobs/127791187.

**Comment by raulchen** [comment]
Comments are all fixed.
BTW, how do I locally run linter for c++ code? just `clang-format changed_files`?

**Review by stephanie-wang — APPROVED** [review]

**Comment by stephanie-wang** [comment]
Thanks! I'll merge once the tests pass.

For linting, in the top-level directory, you can run `$ .travis/git-clang-format`

**Comment by raulchen** [comment]
CI errors seem unrelated

**Pull request merged**